sklearn/cluster/_hierarchical.pyx

```diff
@@ -332,3 +332,93 @@ cdef class WeightedEdge:
                                               self.weight,
                                               self.a, self.b)
 
+
+################################################################################
+# Efficient labelling/conversion of MSTs to single linkage hierarchies
+
+cdef class UnionFind(object):
+
+    cdef ITYPE_t next_label
+    cdef ITYPE_t[:] parent
+    cdef ITYPE_t[:] size
+
+    def __init__(self, N):
+        self.parent = -1 * np.ones(2 * N - 1, dtype=ITYPE, order='C')
+        self.next_label = N
+        self.size = np.hstack((np.ones(N, dtype=ITYPE),
+                               np.zeros(N - 1, dtype=ITYPE)))
+
+    @cython.boundscheck(False)
+    @cython.nonecheck(False)
+    cdef void union(self, ITYPE_t m, ITYPE_t n):
+        self.size[self.next_label] = self.size[m] + self.size[n]
+        self.parent[m] = self.next_label
+        self.parent[n] = self.next_label
+        self.size[self.next_label] = self.size[m] + self.size[n]
+        self.next_label += 1
+
+        return
+
+    @cython.boundscheck(False)
+    @cython.nonecheck(False)
+    cdef ITYPE_t fast_find(self, ITYPE_t n):
+        cdef ITYPE_t p
+        p = n
+        # find the highest node in the linkage graph so far
+        while self.parent[n] != -1:
+            n = self.parent[n]
+        # provide a shortcut up to the highest node
+        while self.parent[p] != n:
+            p, self.parent[p] = self.parent[p], n
+        return n
+
+
+@cython.boundscheck(False)
+@cython.nonecheck(False)
+cpdef np.ndarray[DTYPE_t, ndim=2] single_linkage_label(
+    np.ndarray[DTYPE_t, ndim=2] L):
+    """
+    Convert an linkage array or MST to a tree by labelling clusters at merges.
+    This is done by using a Union find structure to keep track of merges
+    efficiently.
+
+    Parameters
+    ----------
+    L: array of shape (n_samples - 1, 3)
+        The linkage array or MST where each row specifies two samples
+        to be merged and a distance or weight at which the merge occurs. This
+         array is assumed to be sorted by the distance/weight.
+
+    Returns
+    -------
+    A tree in the format used by scipy.cluster.hierarchy.
+    """
+
+    cdef np.ndarray[DTYPE_t, ndim=2] result_arr
+    cdef DTYPE_t[:, ::1] result
+
+    cdef ITYPE_t left, left_cluster, right, right_cluster, index
+    cdef DTYPE_t delta
+
+    result_arr = np.zeros((L.shape[0], 4), dtype=DTYPE)
+    result = (<DTYPE_t[:L.shape[0], :4:1]> (
+        <DTYPE_t *> result_arr.data))
+    U = UnionFind(L.shape[0] + 1)
+
+    for index in range(L.shape[0]):
+
+        left = <ITYPE_t> L[index, 0]
+        right = <ITYPE_t> L[index, 1]
+        delta = L[index, 2]
+
+        left_cluster = U.fast_find(left)
+        right_cluster = U.fast_find(right)
```

Comment:
Here you pass user input to a method that has boundschecks turned off. If the user makes an error, this will lead to a segfault or worse. We'd need to either manually validate all the elements of ``L``, or turn on boundschecking for ``fast_find``

Reply:
That's fair -- ideally this should only really be used internally by code that has suitably constructed ``L``, but there are no guarantees of that and it is an exposed method, so that's definitely an issue. I've opted for turning on bounds checking in ``fast_find``; it should provide relatively little performance hit all things considered (given the runtime complexity at this stage of the algorithm), and is easier than adding validation of L.

Reply:
Is there a good reason to expose the method?

Reply:
Only under the assumption that other people may want to use it at some point. When I look more closely however I see that it is only ever imported into a ``_hierarchical`` namespace, so really people shouldn't see or call it. At that point it is only ever handed data output from ``minimum_spanning_tree`` (suitably sorted) which really ought to be valid, otherwise there are much deeper problems. Is that sufficient control to accept the lack of bounds checking for ``fast_find``?

Reply:
One compromise would be to have a fast cdef'ed version of the method, as well as a public wrapper that does the validation step.

Reply:
It's effectively private now, so I will put the bounds checking off again and simply add a comment that notes that if one wants to reuse the function they should add a validation wrapper. I see no reason to add more code until it becomes clear it is going to actually get used elsewhere.